### ISSUE TITLE
Weird implementation of RectangleF in collision detection

### ISSUE THREAD

**Issue by Acolit** [craftworkgames/MonoGame.Extended#224]
TL;DR: RectangleF uses cartesian coordinates for its boundaries. CollisionGrid uses them as matrix indexes of sorts, which causes problems with grid-based collision detection. **EDIT:** It shouldn't cause any, but somehow, a safeguard right before a call to OnCollision() doesn't seem to work well...

So I ran into an issue yesterday, and spent quite a bit of time trying to figure it out.
Basically, as I was making a simple 2D platformer to get a grasp of MonoGame.Extended's Tiled map integration and collision detection, my player-controlled class' OnCollision() function would be triggered when right next to a wall on the right, but not when that wall was on the left.

After browsing through the source code today, I discovered that it is due to how CollisionGrid.GetCellsOverlappingRectangle() uses the proprieties of the RectangleF struct.

RectangleF is a struct that, according to its summary, "Describes a floating point 2D-rectangle", which is probably why it's using these in a "cartesian values thingamajing" sort of way.

    public float Right => X + Width;
    public float Bottom => Y + Height;

But CollisionGrid.GetCellsOverlappingRectangle() is not treating these values as cartesian coordinates, but as matrix indexes (Because of the CollisionGrid being, well, a grid). It's a bit hard to explain for me, but basically, it interprets the size as having 1 more "pixel" to its width and height, from the top-left corner.

    public IEnumerable<CollisionGridCell> GetCellsOverlappingRectangle(RectangleF rectangle)
        {
            var sx = (int) (rectangle.Left/CellWidth);
            var sy = (int) (rectangle.Top/CellHeight);
            var ex = (int) ((rectangle.Right/CellWidth) + 1);
            var ey = (int) ((rectangle.Bottom/CellHeight) + 1);

            for (var y = sy; y < ey; y++)
            {
                for (var x = sx; x < ex; x++)
                    yield return GetCellAtIndex(x, y);
            }
        }

For example, let's say we have a CollisionWorld object where the collision grid is 3*3 cells in size and each cells are 10*10 pixels (So a stage of 30*30 pixels indexed from 0 to 29). I have an actor about 5*5 pixels in size, which is located at (10, 10). I get:

(int) (10.0f/10) = (int) 1.0f = 1
(int) (10.0f/10) = (int) 1.0f = 1
(int) (((10.0f + 5)/10)+1) = (int) ((15.0f/10)+1) = (int) (1.5f + 1) = 2
(int) (((10.0f + 5)/10)+1) = (int) ((15.0f/10)+1) = (int) (1.5f + 1) = 2

Sounds about right. However, if I instead place it at (15, 15), so right next to another cell on the right and the bottom, I get:

(int) (15.0f/10) = (int) 1.5f = 1
(int) (15.0f/10) = (int) 1.5f = 1
(int) (((15.0f + 5)/10)+1) = (int) ((20.0f/10)+1) = (int) (2.0f + 1) = 3
(int) (((15.0f + 5)/10)+1) = (int) ((20.0f/10)+1) = (int) (2.0f + 1) = 3

While still actually being in the same cell entirely, the second case extends the loop, which then returns more cells than it should.

This means that this function usually works normally, until the actor's BoundingBox gets right next to another cell, in which case instead of returning all cells it intersects with, it returns all cells it intersects with and all the cells directly right or below.

Thankfully it is, in theory, an easy problem to fix, depending of how sprites are rendered.

- If sprites are aligned to the nearest (floored) index in a game's pixel grid, subtracting 1 to the values called by rectangle.Right and rectangle.Bottom right as they are there might do the trick. It doesn't call OnCollision() when the BoundingBox is less than a pixel in though:

    `var ex = (int) (((rectangle.Right - 1)/CellWidth) + 1);`
    `var ey = (int) (((rectangle.Bottom - 1)/CellHeight) + 1);`

- If they are not, or if the bounding box has to keep in place, then conditionally assigning values to ex and ey would work, although with my current knowledge, it would make the code quite a bit heavier, and I can currently only guess whether or not they work:

    `var ex = rectangle.Right%CellWidth == 0 ? (int) (((rectangle.Right - 1)/CellWidth) + 1) : (int) ((rectangle.Right/CellWidth) + 1);`
    `var ey = rectangle.Bottom%CellHeight == 0 ? (int) (((rectangle.Bottom - 1)/CellHeight) + 1) : (int) ((rectangle.Bottom/CellHeight) + 1);`

I'm posting this as an issue rather than just making a commit with the proposed changes, because I am not yet confident enough in my capacity to fix this without creating more issues and because, as I've seen, a rewrite of the implementation of Tiled maps (or at least a rewrite of its renderer) is in the works and I don't want to get ahead of myself.

**EDIT**: So apparently I did get ahead of myself.

The actual place where OnCollision is called has a safeguard:

    foreach (var collidable in _grid.GetCollidables(actor.BoundingBox))
    {
        var intersection = RectangleF.Intersect(collidable.BoundingBox, actor.BoundingBox);

        if (intersection.IsEmpty)
            continue;

        var info = GetCollisionInfo(actor, collidable, intersection);
        actor.OnCollision(info);
    }

So it should be fine tuning itself every time a modification was made to the actor object, ensuring that nothing else triggers OnCollision() if it actually isn't colliding with it.

However, It seems that the problem still exists, and I really don't know at this point...

**Comment by Acolit** [comment]
PS: I'm not sure of how important this is, but I'm using the NuGet package from Visual Studio.

**Comment by craftworkgames** [comment]
Thanks for the detailed bug report :)

So here's the thing. The `CollisionGrid` was kind of written in an afternoon for a demo and was never really finished. To be honest, it was probably a mistake to include it in the library. Shortly after, we decided that it would eventually be replaced by a completely different collision detection system. Some work has been done in that area, but it's not finished.

I've considered removing it from the library and just keeping a copy handy for anyone that still wants to use it. So if you're using the `CollisionGrid` and you're mostly happy with it I think the best option would be to take a copy of the code and include it directly in your game. That way you can customize it your liking.

That said, @LithiumToast might be able to shed some light on the state of the new collision system. It might also be worth reading through #74 and #205.

**Comment by Acolit** [comment]
Thank you for the reply on the matter!

For now, seeing how my goal is to learn how to make a simple platformer with the library, using `CollisionGrid` is currently the easiest way for me to implement collision detection, so I figured after some attempts at making a set of classes for a new one that I would keep it.

The issue was indeed an issue that has been fixed since the last NuGet update of MonoGame.Extended. I may have not installed the newer version properly at first, but I have tried it again recently and it was fixed. I had other issues with collision detection, but I found a workaround to those, and they are unrelated to this particular issue anyway.
